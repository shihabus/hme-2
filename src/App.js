import React, { useRef, useState } from "react";
import axios from "axios";
import PropTypes from "prop-types";
import useStateHook from "./useStateHook";
import validator from "./utils/inputValidator";
import endpoint from "./endpoint";
import initialState from "./initialState";
import { Logo, Text1, Text2, Toast, Modal, SpinnerLoader } from "./components";
import { HEADER, SUB_HEADER } from "./constants/string";
import { LOADING, FAILED, SUCCESS } from "./loginFetchStatus";
import {
  Wrapper,
  Form,
  SubmitButton,
  InputWrapper,
  InputField,
  ErrorText
} from "./styledComponents";

function App({ submit }) {
  const emailRef = useRef(null);
  const passwordRef = useRef(null);
  const [loginStatus, setLoginStatus] = useState(null);

  const [state, setState] = useStateHook(initialState);

  const {
    email,
    password,
    emailError,
    passwordError,
    emailIsValid,
    passwordIsValid
  } = state;

  const handleChange = (type, ref, event) => {
    event.preventDefault();
    const value = event.currentTarget.value;
    const { isValid, errorMessage } = validator(type, value, ref.current);

    setState({
      [`${type}`]: value,
      [`${type}Error`]: errorMessage,
      [`${type}IsValid`]: isValid
    });
  };

  const clear = () => setState(initialState);

  const handleSubmit = async () => {
    try {
      setLoginStatus(LOADING);
      const response = await axios.get(endpoint);
      const data = response.data;
      clear();
      setLoginStatus(SUCCESS);
    } catch (error) {
      setLoginStatus(FAILED);
    }
  };

  return (
    <Wrapper>
      <Form
        aria-label="form"
        onSubmit={e => {
          e.preventDefault();
          submit();
          handleSubmit();
        }}
      >
        <Logo />
        <Text1 txt={HEADER} />
        <Text2 txt={SUB_HEADER} />
        <InputWrapper>
          <InputField
            ref={emailRef}
            type="email"
            aria-label="email"
            value={email}
            onChange={e => handleChange("email", emailRef, e)}
          />
          {!emailIsValid && <ErrorText>{emailError}</ErrorText>}
        </InputWrapper>
        <InputWrapper>
          <InputField
            ref={passwordRef}
            type="password"
            aria-label="password"
            value={password}
            onChange={e => handleChange("password", passwordRef, e)}
          />
          {!passwordIsValid && <ErrorText>{passwordError}</ErrorText>}
        </InputWrapper>
        <SubmitButton
          type="submit"
          value="Login"
          disabled={!emailIsValid || !passwordIsValid}
        />
      </Form>
      {loginStatus === LOADING && (
        <Modal aria-label="loading">
          <SpinnerLoader loadingText="Loading..." />
        </Modal>
      )}
      {loginStatus === SUCCESS && (
        <Toast
          aria-label="success"
          type="success"
          message="Login was successful."
        />
      )}
      {loginStatus === FAILED && (
        <Toast aria-label="failed" type="error" message="Login failed." />
      )}
    </Wrapper>
  );
}

export default App;

App.defaultProps = {
  submit: () => {}
};

App.propTypes = {
  submit: PropTypes.func
};
