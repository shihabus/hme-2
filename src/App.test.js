import React from "react";
import { render, fireEvent, waitForElement, act } from "@testing-library/react";
import "@testing-library/jest-dom/extend-expect";
import App from "./App";
import axiosMock from "axios";

jest.mock("axios");

test("Login form", async () => {
  const mockSubmit = jest.fn();
  const { getByText, getByLabelText } = render(<App submit={mockSubmit} />);

  const form = getByLabelText("form");
  const email = getByLabelText("email");
  const password = getByLabelText("password");
  const submitBtn = getByText(/login/i);

  const inputEmailVal = "fedex@fedex.io";
  fireEvent.change(email, { target: { value: inputEmailVal } });
  expect(email.value).toBe(inputEmailVal);

  const inputPasswordVal = "fakeUser1";
  fireEvent.change(password, { target: { value: inputPasswordVal } });
  expect(password.value).toBe(inputPasswordVal);

  expect(submitBtn.disabled).toBeFalsy();

  axiosMock.get.mockResolvedValueOnce({
    data: {
      result: "success"
    }
  });

  expect(submitBtn).toBeTruthy();
  fireEvent.submit(form);

  await waitForElement(() => {
    expect(getByLabelText("loading")).toBeTruthy();
  });

  await waitForElement(() => {
    expect(getByLabelText("success")).toBeTruthy();
  });

  expect(mockSubmit).toBeCalledTimes(1);
  expect(email.value).toBe("");
  expect(password.value).toBe("");

  expect(submitBtn.disabled).toBeTruthy();
});

test("App snapshot", () => {
  const { container } = render(<App />);
  expect(container.firstChild).toMatchSnapshot();
});
