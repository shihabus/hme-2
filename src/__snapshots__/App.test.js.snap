// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App snapshot 1`] = `
<div
  class="sc-fzozJi hHBylK"
>
  <form
    aria-label="form"
    class="sc-fzoLsD jNtXXC"
  >
    <img
      alt="healthifyme logo"
      class="sc-AxjAm fqmBvb"
      src="test-file-stub"
    />
    <p
      class="sc-AxiKw cCwMHv"
    >
      Sign In
    </p>
    <p
      class="sc-AxiKw sc-AxhCb eINRKv"
    >
      Use you Healthifyme Account
    </p>
    <div
      class="sc-fzplWN izHooe"
    >
      <input
        aria-label="email"
        class="sc-fznyAO cScoOw"
        type="email"
        value=""
      />
      <p
        class="sc-fznKkj fUrdSc"
      />
    </div>
    <div
      class="sc-fzplWN izHooe"
    >
      <input
        aria-label="password"
        class="sc-fznyAO cScoOw"
        type="password"
        value=""
      />
      <p
        class="sc-fznKkj fUrdSc"
      />
    </div>
    <input
      class="sc-fzpans dauElG"
      disabled=""
      type="submit"
      value="Login"
    />
  </form>
</div>
`;
